test/iterator-test.js

```diff
@@ -35,35 +58,250 @@ make('iterator is seekable', function (db, t, done) {
+        t.error(err, 'no error from next()')
+        t.is(ite.cache.length, 0, 'nothing cached')
+        t.same(key.toString(), '4', 'key matches')
+        t.same(value.toString(), '4', 'value matches')
+        ite.end(done)
+      })
+    })
+  })
+})
+
 make('iterator invalid seek', function (db, t, done) {
   var ite = db.iterator()
   ite.seek('zzz')
   ite.next(function (err, key, value) {
-    t.error(err, 'no error')
+    t.error(err, 'no error from next()')
     t.same(key, undefined, 'end of iterator')
     t.same(value, undefined, 'end of iterator')
     ite.end(done)
   })
 })
 
-make('reverse seek from invalid range', function (db, t, done) {
+make('iterator reverse seek from invalid range', function (db, t, done) {
   var ite = db.iterator({reverse: true})
   ite.seek('zzz')
   ite.next(function (err, key, value) {
-    t.error(err, 'no error')
+    t.error(err, 'no error from next()')
     t.same(key.toString(), 'two', 'end of iterator')
     t.same(value.toString(), '2', 'end of iterator')
     ite.end(done)
   })
 })
+
+make('iterator seek lands on or after target', function(db, t, done) {
+  var max = 100, step = 15
+  db.batch(pairs(max, {lex: true, not: even}), function(err){
+    t.error(err, 'no error from batch()')
+
+    var pending = 3
+    var cb = function(err) {
+      t.error(err, 'no error from end()')
+      if (--pending === 0) done()
+    }
+
+    loop(db.iterator(), 0, cb)
+    loop(db.iterator({gt: '0a'}), 10, cb)
+    loop(db.iterator({highWaterMark: 1}), 0, cb)
+
+    function loop(ite, i, done) {
+      if (i >= max) return ite.end(done)
+      ite.seek(lexi.pack(i, 'hex'))
+      ite.next(function (err, key, value) {
+        var expected = even(i) ? i+1 : i
+        t.is(value && value.toString(), ''+expected, 'k'+expected)
+        loop(ite, i+step, done)
+      })
+    }
+  })
+})
+
+make('iterator reverse seek lands on or before target', function(db, t, done) {
+  var max = 100, step = 15
+  db.batch(pairs(max, {lex: true, not: even}), function(err){
+    t.error(err, 'no error from batch()')
+
+    var pending = 3
+    var cb = function(err) {
+      t.error(err, 'no error from end()')
+      if (--pending === 0) done()
+    }
+
+    loop(db.iterator({reverse:true}), max, cb)
+    loop(db.iterator({reverse:true, lt: lexi.pack(90, 'hex')}), 90, cb)
+    loop(db.iterator({reverse:true, highWaterMark: 1}), max, cb)
+
+    function loop(ite, i, done) {
+      if (i <= 0) return ite.end(done)
+      ite.seek(lexi.pack(i, 'hex'))
+      ite.next(function (err, key, value) {
+        var expected = even(i) ? i-1 : i
+        t.is(value && value.toString(), ''+expected, 'k'+expected)
+        loop(ite, i-step, done)
+      })
+    }
+  })
+})
+
+make('iterator seek stays in range', function(db, t, done) {
+  db.batch(pairs(10), function(err){
+    t.error(err, 'no error from batch()')
+
+    var pending = 0
+
+    expect({ gt: 5 }, 5, 6) // hm, should it?
+    expect({ gt: 5 }, 6, 6)
```

Comment:
Given that there's a sequence of numbers in the db, what should this do?

```js
var ite = db.iterator({gt: '5'})
ite.seek('5')
ite.next(function(err, key, value){
  // ?
})
```

- throw, because the seek target (5) is out of range
- yield `undefined`
- yield 6 (because even though the target is out of range, the seek lands on 6, which makes it valid) (this is the current behavior)